rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+easily swap between binaries thus reducing the amount of down-time that nodes
+experience although this is outside the scope of this RFC.
+
+The actual switch would occur in the same manner as an App Version change, using
+the `EndBlockResponse{}` ABCI call to indicate to Tendermint that a different
+block protocol be used. In the case that a node failed to upgrade, the node
+would gracefully stop and error.
+
+Given the delayed execution model, careful consideration would be needed to
+ensure that the nodes could transition smoothly. It may be that the actual
+upgrade might not take place till height: h + 2.
+
+As we expect upgrades to always increment the block version by one, rather
+than having the version number being passed in `EndBlockResponse{}`, another
+option could be to have the height where we want the upgrade to occur. With this
+approach validators are not reaching consensus on the version but the window of
+time (as a height) that nodes will have to perform or prepare for the upgrade.
+
+We will now cover the two main methods of executing a soft upgrade.
+
+## Method 1: Multi Data Structure Support
+
+This would require Tendermint to support all prior block protocol version since
+the last hard fork. This would be done by somehow finding what the version of
+the structure is and processing it accordingly. This mimics a similar design
+model to the Cosmos SDK where all legacy code is stored in the repository and
+conditionals are used to indicate how messages are processed.
+
+One could imagine a directory structure for each module as
+such:
+
+```go
+consensus // module
+|---- state.go
+|---- state_test.go
+|       ...
+|---- votes.go // current version (v10)
+|       ...
+|---- legacy
+   |---- v8 // prior supported versions
+      |---- votes.go
+      |      ...    // other data structures ( and pure functions )
+   |---- v9
+      |---- votes.go
+      |      ...
+```
+
+The module, depending on the types, may statically cast each of the different
+versions or define an interface that it expects all versions to follow. For
+example:
+
+```go
+func (blockExec *BlockExecutor) CreateProposalBlock(state State) (BlockI, error)
+{
+ switch state.Version.Block {
+ case 9:
+  return v9.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ case 10:
+  return v10.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ default:
+  return fmt.Errorf("block version %d not recognised", state.Version.Block)
+ }
+}
+```
+
+In this case, Block which falls under a data structure of the block protocol
+could have many versions and is represented here as an interface. `State` and
+`BlockExecutor` are structs outside of the block protocol so would be
+internalized to allow for changes (so long as this wouldn't change the overall
+behavior).
+
+Another similar approach is to have the prior versions in a different repo and
+import it as a library. This may minimize the amount of code in the repo.
```

Comment:
This sounds like it may lead to dependency hell, but if this were a priority, we could structure our codebase to support it.